### ISSUE TITLE
registering block devices after starting the kernel

### ISSUE THREAD

**Issue by M1cha** [lkl/linux#235]
if I do that I just get a 'can't mount disk: No such device or address' which actually comes from 'get_virtio_blkdev' which can't find the sys node for the new device.

The root problem here probably is that the virtio-mmio device doesn't get registered.

**Comment by thehajime** [comment]
currently virtio-mmio code is created/instantiated by boot cmdline parameters so this case (i.e., creating after `lkl_start_kernel`) is not supported.

We may introduce other ways to instantiate it - would be another PR.

https://github.com/lkl/linux/blob/master/drivers/virtio/virtio_mmio.c#L9

**Comment by M1cha** [comment]
@thehajime I'm currently implementing that but I'm not sure what's the correct way to define a new syscall(it can't be called directly because adding devices causes a resched of the idle thread otherwise).

there already is a __NR_create_syscall_thread which is not used anywhere, also the function sys_create_syscall_thread is defined manually without the SYCALL_DEFINE wrapper and then called directly without any syscalls.

I just used SYSCALL_DEFINE to define my syscall function, added the declaration in unistd.h and the function gets correctly inserted into the syscall table and in the generated header  'syscall_defs.h' is a declaration too, but it's wrapped with '#ifdef __NR_virtio_mmio_device_add' which is not defined in the public unistd.h header.

any ideas?

**Comment by tavip** [comment]
@M1cha you can define it in arch/lkl/include/uapi/asm/unistd.h like

#define __NR_virtio_mmio_device_add      __NR_arch_specific_syscall+1

**Comment by M1cha** [comment]
do you have any ideas about a better way to find the /sys node of a block device?
it looks like the virtio device ID is not related to the drive letter at all.

**Comment by tavip** [comment]
Hmm, there should be a perfect correlation between the block device letter and the id returned by lkl_disk_add unless you are using multiple threads.

Can you give more information about the issue you are running into?

**Comment by M1cha** [comment]
@tavip that's just with the current code adds devices in the same order as they are counted by lkl and devices never get removed.

If I set the id of the first mmio device to 5 it will still be a and not e.
Also when starting to remove devices things will start to look even more different.

**Comment by tavip** [comment]
@M1cha Ah, I see. I assume you are using platform_device_add to add device dynamically after starting the kernel? In that case you can try to use /sysfs/devices/platform/virtio-mmio.X/virtioY/block/vdZ/dev to get the major and minor.

The problem is that while X should be the device id you set, Y and Z might get assign independently. So you probably need to use getdents in  /sysfs/devices/platform/virtio-mmio.X/ to look for a directory starting with virtio and then the same for vdZ...

BTW, I think it is better to use the auto assignment for platform devs, i.e. use PLATFORM_DEVID_AUTO for the platform device id.

**Comment by M1cha** [comment]
@tavip thx I'm actually using PLATFORM_DEVID_AUTO already. the example with '5' was just to clarify the problem.

Shouldn't Y and Z always be 0 if the device doesn't contain a partition table?

**Comment by tavip** [comment]
@M1cha No, Y is the id of the virtio device (drivers/virtio/virtio.c), and now that I think about it, it should be equal with X because we should not have other way of spawning a virtio-device (see bellow) other then virtio-mmio.

The full device hierarcy is virtio-mmio-device -> virtio-device -> virtio-block-device but each device has its own id.

So X and Y should always be the same for LKL. Z might be different because we might have both blk and net devices and we might end up with the following hierarchy:

virtio-mmio.0 -> virtio.0 -> vda
virtio-mmio.1 -> virtio.1 -> netdev0
virtio-mmio.2 -> virtio.2 -> vdb

**Comment by M1cha** [comment]
@tavip there seems to be another problem with the number of id's.
The virtio-block-device assigns minor id's with wholes of 16(0-16-32-...256).
If I use the kernel's internal macros to create the block device then(the uapi ones only support id's up to 256), that works perfectly(I verified it using stat) - unless I try to mount such a device which results in "No such device or address"

**Comment by tavip** [comment]
@M1cha Yes, it looks like we are using the wrong macros. But I don't think internal macros work either, see new_decode_dev. It looks like we need to use something like this (picked from glibc):

```
__extension__ __extern_inline __attribute_const__ unsigned long long int
__NTH (gnu_dev_makedev (unsigned int __major, unsigned int __minor))
{
  return ((__minor & 0xff) | ((__major & 0xfff) << 8)
          | (((unsigned long long int) (__minor & ~0xff)) << 12)
          | (((unsigned long long int) (__major & ~0xfff)) << 32));
}
```

Also, note that once we skip past ~26 devices, the block device names will start to be formated like vdXY, e.g. vdaw.

**Comment by M1cha** [comment]
@tavip I know, what confuses me is that even after using the correct macros and verifying the created block device node using stat, I can't mount any device with a minor number bigger than 256.

**Comment by tavip** [comment]
@M1cha Hmm, with the above makdev function I was able to mount.

I've pushed my hacks here: https://github.com/tavip/linux/commits/tmp and used lklfuse to test. go to tools/lkl/tests and run lklfuse.sh which should mount somewhere in tmp where you can inspect the full fs namespace as seen by the lkl instance.

**Comment by M1cha** [comment]
@tavip the problem seems to be the macro's I've used. Where do you have makedev from?
I used the kernel's internal macro's

```
#define MINORBITS	20
#define MINORMASK	((1U << MINORBITS) - 1)

#define MAJOR(dev)	((unsigned int) ((dev) >> MINORBITS))
#define MINOR(dev)	((unsigned int) ((dev) & MINORMASK))
#define MKDEV(ma,mi)	(((ma) << MINORBITS) | (mi))
```

**Comment by tavip** [comment]
@M1cha I've picked the version I've reference from glibc - see man makedev. I am not sure why it uses 64bit, it does not seems to be used yet in kernel. There is a huge_decode_dev which right now is just calling new_decode_dev, and I assume it for future expansion.

**Comment by M1cha** [comment]
@tavip I'll make a PR as soon as #239 is merged because I based on top of that commit and we'll have to PR's with the same commit otherwise.

while speaking about limits:
```NR_IRQS                        ((int)sizeof(long) * 8)```
is there a reason for the relationship to the size of long? while I get 64 IRQ's on a 64bit machine, on ARM(32bit) I only get 32 IRQ's which I can end up using if I register too many mmio devices.

**Comment by tavip** [comment]
@M1cha I've just merged #239, thanks for reminding me about it ! It looks like GitHUB doesn't send notifications when new versions are pushed :(

With regard to irq limits, the limitation comes from the various __sync_* functions we use in lkl_trigger_irq() and run_irqs() that can only operate with longs AFAIR.

But I think the synchronization will properly work if we extend it to use multiple such longs. I'll work on something and push a PR soon.

**Comment by M1cha** [comment]
this got fixed with #241
